ocean_lib/assets/v4/asset.py

```diff
@@ -0,0 +1,264 @@
+
+logger = logging.getLogger("ddo")
+
+
+@enforce_types
+class V4Asset:
+    """Asset class to create, import, export, validate Asset/DDO objects for V4."""
+
+    def __init__(
+        self,
+        did: Optional[str] = None,
+        context: Optional[list] = None,
+        chain_id: Optional[int] = None,
+        metadata: Optional[dict] = None,
+        services: Optional[list] = None,
+        credentials: Optional[dict] = None,
+        nft: Optional[dict] = None,
+        datatokens: Optional[list] = None,
+        event: Optional[dict] = None,
+        stats: Optional[dict] = None,
+    ) -> None:
+
+        self.did = did
+        self.context = context or ["https://w3id.org/did/v1"]
+        self.chain_id = chain_id
+        self.metadata = metadata
+        self.version = "4.0.0"
+        self.services = services or []
+        self.credentials = credentials or {}
+        self.nft = nft
+        self.datatokens = datatokens
+        self.event = event
+        self.stats = stats
+
+    @property
+    def requires_address_credential(self) -> bool:
+        """Checks if an address credential is required on this asset."""
+        manager = AddressCredential(self)
+        return manager.requires_credential()
+
+    @property
+    def allowed_addresses(self) -> list:
+        """Lists addresses that are explicitly allowed in credentials."""
+        manager = AddressCredential(self)
+        return manager.get_addresses_of_class("allow")
+
+    @property
+    def denied_addresses(self) -> list:
+        """Lists addresses that are explicitly denied in credentials."""
+        manager = AddressCredential(self)
+        return manager.get_addresses_of_class("deny")
+
+    def add_address_to_allow_list(self, address: str) -> None:
+        """Adds an address to allowed addresses list."""
+        manager = AddressCredential(self)
+        manager.add_address_to_access_class(address, "allow")
+
+    def add_address_to_deny_list(self, address: str) -> None:
+        """Adds an address to the denied addresses list."""
+        manager = AddressCredential(self)
+        manager.add_address_to_access_class(address, "deny")
+
+    def remove_address_from_allow_list(self, address: str) -> None:
+        """Removes address from allow list (if it exists)."""
+        manager = AddressCredential(self)
+        manager.remove_address_from_access_class(address, "allow")
+
+    def remove_address_from_deny_list(self, address: str) -> None:
+        """Removes address from deny list (if it exists)."""
+        manager = AddressCredential(self)
+        manager.remove_address_from_access_class(address, "deny")
+
+    @classmethod
+    def from_dict(cls, dictionary: dict) -> "V4Asset":
+        """Import a JSON dict into this Asset."""
+        values = copy.deepcopy(dictionary)
+        id_key = "id" if "id" in values else "_id"
+        if "services" in values:
+            services = [V4Service.from_dict(value) for value in values.pop("services")]
```

Comment:
or a one-liner: services = [] if "services" not in values else [V4Service.from_dict(value) for value in values.pop("services")]